ddtrace/ext/git.py

```diff
@@ -13,3 +28,88 @@
+COMMIT_AUTHOR_EMAIL = "git.commit.author.email"
+
+# Git Commit Author Date (UTC)
+COMMIT_AUTHOR_DATE = "git.commit.author.date"
+
+# Git Commit Committer Name
+COMMIT_COMMITTER_NAME = "git.commit.committer.name"
+
+# Git Commit Commiter Email
+COMMIT_COMMITTER_EMAIL = "git.commit.committer.email"
+
+# Git Commit Committer Date (UTC)
+COMMIT_COMMITTER_DATE = "git.commit.committer.date"
+
+# Git Commit Message
+COMMIT_MESSAGE = "git.commit.message"
+
+log = get_logger(__name__)
+
+
+def extract_user_info(author=True):
+    # type: (bool) -> Tuple[str, str, str]
+    """Extract git commit author/committer info."""
+    # Note: `git show -s --format... --date...` is suported since git 2.1.4 onwards
+    formatting = "--format=%an,%ae,%ad" if author else "--format=%cn,%ce,%cd"
+    cmd = subprocess.Popen(
+        ["git", "show", "-s", formatting, "--date=format:%Y-%m-%dT%H:%M:%S%z"],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    stdout, stderr = cmd.communicate()
+    if cmd.returncode == 0:
+        name, email, date = compat.ensure_text(stdout).strip().split(",")
+        return name, email, date
+    raise ValueError(stderr)
+
+
+def extract_repository_url():
+    # type: () -> str
+    # Note: `git show ls-remote --get-url` is suported since git 2.6.7 onwards
+    cmd = subprocess.Popen(["git", "ls-remote", "--get-url"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = cmd.communicate()
+    if cmd.returncode == 0:
+        repository_url = compat.ensure_text(stdout).strip()
+        return repository_url
+    raise ValueError(stderr)
+
+
+def extract_commit_message():
+    # type: () -> str
+    # Note: `git show -s --format... --date...` is suported since git 2.1.4 onwards
+    cmd = subprocess.Popen(["git", "show", "-s", "--format=%s"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = cmd.communicate()
+    if cmd.returncode == 0:
+        commit_message = compat.ensure_text(stdout).strip()
+        return commit_message
+    raise ValueError(stderr)
+
+
+def extract_git_metadata():
+    # type: () -> Dict[str, Optional[str]]
+    """Extract git commit metadata."""
+    tags = {}  # type: Dict[str, Optional[str]]
+    try:
+        tags[REPOSITORY_URL] = extract_repository_url()
+        tags[COMMIT_MESSAGE] = extract_commit_message()
+        author = extract_user_info(author=True)
+        tags[COMMIT_AUTHOR_NAME] = author[0]
+        tags[COMMIT_AUTHOR_EMAIL] = author[1]
+        tags[COMMIT_AUTHOR_DATE] = author[2]
+        committer = extract_user_info(author=False)
+        tags[COMMIT_COMMITTER_NAME] = committer[0]
+        tags[COMMIT_COMMITTER_EMAIL] = committer[1]
+        tags[COMMIT_COMMITTER_DATE] = committer[2]
+    except GitNotFoundError:
+        log.error("Git executable not found, cannot extract git metadata.")
+    except ValueError:
+        log.error("Error extracting git metadata, received non-zero return code.", exc_info=True)
+
+    return tags
```

Comment:
we are ok returning a subset of values based on which extract functions succeed or not?

if we want to aim for higher level of completeness, you might want to wrap each `extract_` in a try/except?

for example, if `extract_repository_url` fails, but `extract_user_info` and `extract_commit_message` would succeed, then you would end up extracting no data at all.

Reply:
This is a tricky subject, but from the git commands used by each `extract_` method, 2 use:
```
git show -s --format=...
```
and `extract_repository_url` uses `git ls-remote ...`. My gut says if one fails, the others should fail as well and vice versa.

Then again, we don't really lose anything by wrapping each `extract_` in try/except except for some reduced readability, but I'm open to making that change. WDYT?

Reply:
> My gut says if one fails, the others should fail as well and vice versa.

then all good, if that is the expectation from the code, then this code is good as-is 👍🏻 thanks for talking through it.